lib/filterx/func-str.c

```diff
@@ -0,0 +1,512 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "filterx/func-str.h"
+#include "filterx/object-primitive.h"
+#include "filterx/filterx-object.h"
+#include "filterx/expr-literal.h"
+#include "filterx/expr-literal-generator.h"
+#include "filterx/filterx-eval.h"
+#include "object-extractor.h"
+#include "object-string.h"
+#include "filterx/object-json.h"
+#include "filterx/object-list-interface.h"
+
+#define FILTERX_FUNC_STARTSWITH_USAGE "Usage: startswith(string, prefix, ignorecase=true)" \
+"or startswith(string, [prefix_1, prefix_2, ..], ignorecase=true)"
+
+#define FILTERX_FUNC_ENDSWITH_USAGE "Usage: endswith(string, suffix, ignorecase=true)" \
+"or endswith(string, [suffix_1, suffix_2, ..], ignorecase=true)"
+
+#define FILTERX_FUNC_INCLUDES_USAGE "Usage: includes(string, substring, ignorecase=true)" \
+"or includes(string, [substring_1, substring_2, ..], ignorecase=true)"
+
+typedef struct _FilterXStringWithCache
+{
+  FilterXExpr *expr;
+  FilterXObject *obj;
+  const gchar *borrowed_str_value;
+  gchar *owned_str_value;
+  gsize str_len;
+} FilterXStringWithCache;
+
+typedef gboolean (*FilterXExprAffixProcessFunc)(const gchar *haystack, gsize haystack_len, const gchar *needle,
+                                                gsize needle_len);
+typedef struct _FilterXExprAffix
+{
+  FilterXFunction super;
+  gboolean ignore_case;
+  FilterXExpr *haystack;
+  struct
+  {
+    FilterXExpr *expr;
+    GPtrArray *objects_with_cache;
+  } needle;
+  FilterXExprAffixProcessFunc process;
+} FilterXExprAffix;
+
+static gboolean
+filterx_string_with_cache_is_const(FilterXStringWithCache *self)
+{
+  return self->expr && filterx_expr_is_literal(self->expr);
+}
+
+static gboolean
+filterx_string_with_cache_has_cache(FilterXStringWithCache *self)
+{
+  return self->borrowed_str_value != NULL || self->owned_str_value != NULL;
+}
+
+static void
+filterx_string_with_cache_deinit(FilterXStringWithCache *self)
+{
+  g_free(self->owned_str_value);
+  self->owned_str_value = NULL;
+  self->borrowed_str_value = NULL;
+  self->str_len = 0;
+  filterx_object_unref(self->obj);
+  self->obj = NULL;
+}
+
+static void
+filterx_string_with_cache_free(FilterXStringWithCache *self)
+{
+  g_free(self->owned_str_value);
+  if (self->obj)
+    filterx_object_unref(self->obj);
+  if(self->expr)
+    filterx_expr_unref(self->expr);
+  g_free(self);
+}
+
+static gboolean
+filterx_string_with_cache_do_casefold(FilterXStringWithCache *self)
+{
+  if(self->str_len > G_MAXSSIZE)
+    return FALSE;
+
+  self->owned_str_value = g_utf8_casefold(self->borrowed_str_value, (gssize) self->str_len);
```

Comment:
The string comparisons might need `g_utf8_normalize()` as well besides case insensitive form.
I did not use it yet as I wanted a confirmation about the UTF8-ness of strings in this context.
@bazsi @alltilla  What do you think? :eyes:

Reply:
The official docs say for the deprecated [`strncasecmp`](https://docs.gtk.org/glib/func.strncasecmp.html):
> [...] There are therefore two replacement techniques:
> [g_ascii_strncasecmp()](https://docs.gtk.org/glib/func.ascii_strncasecmp.html), which only works on ASCII and is not locale-sensitive,
> 
> and [**g_utf8_casefold**()](https://docs.gtk.org/glib/func.utf8_casefold.html) **followed by strcmp() on the resulting strings, which is good for case-insensitive sorting of UTF-8.**